### ISSUE TITLE
Any recommended recipes for hosting a remote imaging data server such as the viewer's default?

### ISSUE THREAD

**Issue by omartoutounji** [OHIF/Viewers#868]
By default, the viewer is configured to connect to a remote server hosted by dcmjs.org. 

I would like to duplicate such a server for my own medical imaging data, any instructions/recipes/resources on how to do so? I am mainly looking for ease of use and simple integration with the viewer.

Some of you might recommend Google Cloud Healthcare, and I definitely agree with you as I've been using it for a while for local development purposes and it's great for that. However, I recently realized that it is not so ideal for deployment purposes, if you have time constraints such as I do. 

Why? If you want to associate your own domain with your Google Cloud Healthcare dataset, it has to be approved and verified by Google first, this process can take anywhere from a week to 8 weeks. I can't afford waiting that long, given my time constraints. I also noticed their process isn't transparent at all so I'm just waiting blindly here.

Has anyone gone through this issue with Google Cloud Healthcare? I would appreciate any tips/hacks I can use to speed this up or guidance on how to do it the right way (if I'm doing it the wrong way, I'm not even sure).

**TL;DR:** any way I can duplicate the viewer's default data source for my own data? Has anyone tried to get their Google Cloud Healthcare dataset associated with their own domain (not localhost)?

**Comment by swederik** [comment]
Hi @omartoutounji,

So that deployment is:
- Keycloak for an identity provider
- Dcm4chee-arc-light as a PACS
- Let's Encrypt for SSL
- Kubernetes nginx Ingress as a reverse proxy
- Kubernetes for deployment
- Hosted on AWS Lightsail

I really would like to make this into a one-click setup system but honestly it's still rather complicated. dcm4chee has a pretty good description for how to get this set up on their wiki: https://github.com/dcm4che/dcm4chee-arc-light/wiki/Run-secured-archive-services-on-a-single-host . We had started a repository to try to make easy-to-use examples for how to get this stuff running: https://github.com/OHIF/integration-examples/tree/master/ohif-dcm4chee-secureUI but because of our limited bandwidth it sort of fell down the priority list.

There is some pieces for the Kubernetes deployment here: https://bitbucket.org/Jonaven/namic/src/nginx-ingress/

The biggest problem we ran into is that it's very hard for someone to set this same system up on a local machine e.g. via minikube. This is also the deployment approach for https://crowds-cure.org.

Here are my basic notes, which is the start of a README for that repo. This stuff gets out of date so quickly though (e.g. Calico version is already out of date):

# Rough README for how this stuff is deployed:
**I am pretty sure this stuff is now not working, but maybe it can help somehow.**

# Setup on an Ubuntu VM

1. Setup a virtual machine with a static IP address.

	- Record your public IP: X.X.X.X
	- Record your private IP: Y.Y.Y.Y

2. [Install kubeadm and its dependencies](https://kubernetes.io/docs/setup/independent/install-kubeadm/)

	- Open ports 443, 6443, 2379 - 2380, 10250 - 10252 to the external world. **Note:** This is critical. If you don't do this, the `kubeadm init` command will fail.

	````bash
	PUBLIC_IP=$(curl -s http://whatismyip.akamai.com/)
	sudo locale-gen en_US.UTF-8
	sudo swapoff -a
	sudo apt-get update
	
	= INSTALL DOCKER
	= Start docker service via systemctl

	sudo apt-get install -y apt-transport-https curl
	sudo curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key add -
	sudo touch /etc/apt/sources.list.d/kubernetes.list
	sudo bash -c 'echo "deb http://apt.kubernetes.io/ kubernetes-xenial main" >> /etc/apt/sources.list.d/kubernetes.list'
	sudo apt-get update
	sudo apt-get install -y kubelet kubeadm kubectl
	sudo apt-mark hold kubelet kubeadm kubectl
	sudo kubeadm config images pull
	# sudo kubeadm init --pod-network-cidr=192.168.0.0/16 --apiserver-advertise-address=${PUBLIC_IP}
	sudo kubeadm init --pod-network-cidr=192.168.0.0/16
	mkdir -p $HOME/.kube
	sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
	sudo chown $(id -u):$(id -g) $HOME/.kube/config
	echo "export KUBECONFIG=$HOME/.kube/config" | tee -a ~/.bashrc
	````

4. Install Calico

	````bash
	kubectl apply -f \
	https://docs.projectcalico.org/v3.2/getting-started/kubernetes/installation/hosted/etcd.yaml
	kubectl apply -f \
	https://docs.projectcalico.org/v3.2/getting-started/kubernetes/installation/rbac.yaml
	kubectl apply -f \
	https://docs.projectcalico.org/v3.2/getting-started/kubernetes/installation/hosted/calico.yaml
	````

5. Taint your master node so that it can run Pods
	- Note: You only want to do this if you are running a single-node system

	````bash
	kubectl taint nodes --all node-role.kubernetes.io/master-
	````


6. Clone the repository

	````bash
	git clone https://bitbucket.org/Jonaven/namic.git
	cd namic
	git checkout minikube
	````


7. Deploy the Kubernetes dashboard:

	````bash
	kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/master/src/deploy/recommended/kubernetes-dashboard.yaml
	kubectl apply -f ./optional/kube-dashboard-access.yml
	````

8. Copy the kubectl config to your local development machine

	````bash
	# Run this on your local machine
	scp ubuntu@52.47.104.159:/home/ubuntu/.kube/config .

	# If you have kubectl installed locally,
	# you can now use the following to load
	# the Kubernetes dashboard:

	kubectl --kubeconfig config proxy
	````


	- Get the dashboard token you will use for login
	````bash
	kubectl -n kube-system describe secret $(kubectl -n kube-system get secret | awk '/^kubernetes-dashboard-token-/{print $1}') | awk '$1=="token:"{print $2}'
	````

	The dashboard is at the URL:
	http://127.0.0.1:8001/api/v1/namespaces/kube-system/services/https:kubernetes-dashboard:/proxy/#!/pod?namespace=default

9. Setup the Nginx Ingress Mandatory:


	````bash
	kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/mandatory.yaml
	````


10. Start the main services

	````bash
	kubectl apply -f ./	
	````

11. Set your private (*not public!*) IP and add it as an externalIP for the Ingress Service (ingress-nginx.yml)

 (e.g. 172.26.12.142)

 	````bash
 	sed -i 's/192.168.99.100/172.26.12.142/g' ingress-nginx.yml;
 	kubectl apply -f ./ingress-nginx.yml
 	````

12. Ensure your DNS points to your server for the hostname you want to use

	- You will need an A record pointing from your domain to your public IP. In our case, this is:

	cancer 107.20.60.149

13. Switch all IP addresses to your domain

	````bash
	find . -name '*.yml' -exec sed -i -e 's/http:\/\/192.168.99.100/https:\/\/cancer.crowds-cure.org/g' {} \;
	kubectl apply -f ./
	````

	````bash
	# Update secured/ingress.yml to include your domain
	kubectl apply -f secured/ingress.yml
	````

14. Uncomment the relevant portions of the nginx configuration (TODO: Make two copies of the Nginx config?)

15. Run the Lets Encrypt job

	- ** First, verify you can reach your site! **

	````bash
	# This must not return a 404
	curl -H "HOST: k8s-testing.ohif.org" http://k8s-testing.ohif.org/
	````

	- Switch optional/letsencrypt DOMAIN to match your site

	````bash
	sed -i 's/kubernetes-example.ohif.org/k8s-testing.ohif.org/g' ./optional/letsencrypt.yml;
	````
	
	
	````bash
	kubectl apply -f ./optional/letsencrypt.yml
	kubectl apply -f ./nginx.yml
	````

16. Edit Keycloak to force SSL redirection

**Note**: Edit keycloak.xml, not standalone or standalone-ha

https://www.keycloak.org/docs/latest/server_installation/index.html#_setting-up-a-load-balancer-or-proxy

https://www.keycloak.org/docs/latest/server_installation/index.html#enable-https-ssl-with-a-reverse-proxy

	````bash
	sudo nano /data/dcm4chee-arc/keycloak/configuration/keycloak.xml
	````

17. Add the Lets Encrypt certificate to the Keycloak Truststore

	- Retrieve the certificate

	````bash
	kubectl get secret letsencrypt-certs -o yaml
	````

	- Copy the certificate (tls.crt) and decode it from base64:

	````bash
	echo 'Paste tls.crt here' | base64 --decode > certificate.crt
	````

	````bash
	sudo apt install openjdk-9-jre-headless

	sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias letsencrypt -file certificate.crt

	sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias letsencrypt -file certificate.crt

	- Remove the decoded certificate, for safety:

		````bash
		rm certificate.crt
		````

	- Note: If you are using Auth0, you will need to add the AmazonRootCA1 certificate as well! See https://www.amazontrust.com/repository/AmazonRootCA1.pem

	- If you are using Auth0 with a custom domain, you will need:

	````
	 curl https://letsencrypt.org/certs/isrgrootx1.pem.txt > isrgrootx1.pem
	 curl https://letsencrypt.org/certs/lets-encrypt-x3-cross-signed.pem.txt > lets-encrypt-x3-cross-signed.pem
	 curl https://letsencrypt.org/certs/letsencryptauthorityx3.pem.txt > letsencryptauthorityx3.pem

	 sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias le1 -file isrgrootx1.pem
	 sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias le2 -file lets-encrypt-x3-cross-signed.pem
	 sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias le3 -file letsencryptauthorityx3.pem

	 sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias le1 -file isrgrootx1.pem
	 sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias le2 -file lets-encrypt-x3-cross-signed.pem
	 sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias le3 -file letsencryptauthorityx3.pem
	````

	https://groups.google.com/forum/#!searchin/dcm4che/Let$27s$20Encrypt%7Csort:date/dcm4che/rtMYHztMWxY/chINDE4RCAAJ


19. Create OIDC clients for dcm4chee and the OHIF Viewer in Keycloak

**Note:** You will need to add a Valid Redirect URI for the HTTP route to dcm4chee, in addition to the HTTPs route. This is a known issue.

20. (Optional) Setup IHE Invoke Image Display routes in dcm4chee

Notes:

- Had to disable SSL to set up Keycloak initially
- If LE certificate needs to be renewed, you may have to add it to dcm4chee and keycloak's keystore again
- If you run into 'no path to resource' errors it's probably an issue like "part of the SSL cert is missing in the keystore"

**Comment by omartoutounji** [comment]
> Hi @omartoutounji,
> 
> So that deployment is:
> 
> * Keycloak for an identity provider
> * Dcm4chee-arc-light as a PACS
> * Let's Encrypt for SSL
> * Kubernetes nginx Ingress as a reverse proxy
> * Kubernetes for deployment
> * Hosted on AWS Lightsail
> 
> I really would like to make this into a one-click setup system but honestly it's still rather complicated. dcm4chee has a pretty good description for how to get this set up on their wiki: https://github.com/dcm4che/dcm4chee-arc-light/wiki/Run-secured-archive-services-on-a-single-host . We had started a repository to try to make easy-to-use examples for how to get this stuff running: https://github.com/OHIF/integration-examples/tree/master/ohif-dcm4chee-secureUI but because of our limited bandwidth it sort of fell down the priority list.
> 
> There are some pieces for the Kubernetes deployment here: https://bitbucket.org/Jonaven/namic/src/nginx-ingress/
> 
> The biggest problem we ran into is that it's very hard for someone to set this same system up on a local machine e.g. via minikube. This is also the deployment approach for https://crowds-cure.org.
> 
> If there is someone in the community that would like to help setting up e.g. a Helm chart to get all this stuff working easily for end users, we would appreciate the help.
> 
> Here are my basic notes, which is the start of a README for that repo. This stuff gets out of date so quickly though (e.g. Calico version is already out of date):
> 
> # Rough README for how this stuff is deployed:
> **I am pretty sure this stuff is now not working, but maybe it can help somehow.**
> 
> # Setup on an Ubuntu VM
> 1. Setup a virtual machine with a static IP address.
>    
>    * Record your public IP: X.X.X.X
>    * Record your private IP: Y.Y.Y.Y
> 2. [Install kubeadm and its dependencies](https://kubernetes.io/docs/setup/independent/install-kubeadm/)
>    
>    * Open ports 443, 6443, 2379 - 2380, 10250 - 10252 to the external world. **Note:** This is critical. If you don't do this, the `kubeadm init` command will fail.
>    
>    ```shell
>    PUBLIC_IP=$(curl -s http://whatismyip.akamai.com/)
>    sudo locale-gen en_US.UTF-8
>    sudo swapoff -a
>    sudo apt-get update
>    
>    = INSTALL DOCKER
>    = Start docker service via systemctl
>    
>    sudo apt-get install -y apt-transport-https curl
>    sudo curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key add -
>    sudo touch /etc/apt/sources.list.d/kubernetes.list
>    sudo bash -c 'echo "deb http://apt.kubernetes.io/ kubernetes-xenial main" >> /etc/apt/sources.list.d/kubernetes.list'
>    sudo apt-get update
>    sudo apt-get install -y kubelet kubeadm kubectl
>    sudo apt-mark hold kubelet kubeadm kubectl
>    sudo kubeadm config images pull
>    # sudo kubeadm init --pod-network-cidr=192.168.0.0/16 --apiserver-advertise-address=${PUBLIC_IP}
>    sudo kubeadm init --pod-network-cidr=192.168.0.0/16
>    mkdir -p $HOME/.kube
>    sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
>    sudo chown $(id -u):$(id -g) $HOME/.kube/config
>    echo "export KUBECONFIG=$HOME/.kube/config" | tee -a ~/.bashrc
>    ```
> 3. Install Calico
>    ```shell
>    kubectl apply -f \
>    https://docs.projectcalico.org/v3.2/getting-started/kubernetes/installation/hosted/etcd.yaml
>    kubectl apply -f \
>    https://docs.projectcalico.org/v3.2/getting-started/kubernetes/installation/rbac.yaml
>    kubectl apply -f \
>    https://docs.projectcalico.org/v3.2/getting-started/kubernetes/installation/hosted/calico.yaml
>    ```
> 4. Taint your master node so that it can run Pods
>    
>    * Note: You only want to do this if you are running a single-node system
>    
>    ```shell
>    kubectl taint nodes --all node-role.kubernetes.io/master-
>    ```
> 5. Clone the repository
>    ```shell
>    git clone https://bitbucket.org/Jonaven/namic.git
>    cd namic
>    git checkout minikube
>    ```
> 6. Deploy the Kubernetes dashboard:
>    ```shell
>    kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/master/src/deploy/recommended/kubernetes-dashboard.yaml
>    kubectl apply -f ./optional/kube-dashboard-access.yml
>    ```
> 7. Copy the kubectl config to your local development machine
>    ```shell
>    # Run this on your local machine
>    scp ubuntu@52.47.104.159:/home/ubuntu/.kube/config .
>    
>    # If you have kubectl installed locally,
>    # you can now use the following to load
>    # the Kubernetes dashboard:
>    
>    kubectl --kubeconfig config proxy
>    ```
>    
>    
>    
>    * Get the dashboard token you will use for login
>    
>    ```shell
>    kubectl -n kube-system describe secret $(kubectl -n kube-system get secret | awk '/^kubernetes-dashboard-token-/{print $1}') | awk '$1=="token:"{print $2}'
>    ```
>    
>    
>    The dashboard is at the URL:
>    http://127.0.0.1:8001/api/v1/namespaces/kube-system/services/https:kubernetes-dashboard:/proxy/#!/pod?namespace=default
> 8. Setup the Nginx Ingress Mandatory:
>    ```shell
>    kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/mandatory.yaml
>    ```
> 9. Start the main services
>    ```shell
>    kubectl apply -f ./	
>    ```
> 10. Set your private (_not public!_) IP and add it as an externalIP for the Ingress Service (ingress-nginx.yml)
> 
> (e.g. 172.26.12.142)
> 
> ```
> ````bash
> sed -i 's/192.168.99.100/172.26.12.142/g' ingress-nginx.yml;
> kubectl apply -f ./ingress-nginx.yml
> ````
> ```
> 
> 1. Ensure your DNS points to your server for the hostname you want to use
>    
>    * You will need an A record pointing from your domain to your public IP. In our case, this is:
>    
>    cancer 107.20.60.149
> 2. Switch all IP addresses to your domain
>    ```shell
>    find . -name '*.yml' -exec sed -i -e 's/http:\/\/192.168.99.100/https:\/\/cancer.crowds-cure.org/g' {} \;
>    kubectl apply -f ./
>    ```
>    
>    
>    ```shell
>    # Update secured/ingress.yml to include your domain
>    kubectl apply -f secured/ingress.yml
>    ```
> 3. Uncomment the relevant portions of the nginx configuration (TODO: Make two copies of the Nginx config?)
> 4. Run the Lets Encrypt job
>    
>    * ** First, verify you can reach your site! **
>    
>    ```shell
>    # This must not return a 404
>    curl -H "HOST: k8s-testing.ohif.org" http://k8s-testing.ohif.org/
>    ```
>    
>    
>    
>    * Switch optional/letsencrypt DOMAIN to match your site
>    
>    ```shell
>    sed -i 's/kubernetes-example.ohif.org/k8s-testing.ohif.org/g' ./optional/letsencrypt.yml;
>    ```
>    
>    
>    ```shell
>    kubectl apply -f ./optional/letsencrypt.yml
>    kubectl apply -f ./nginx.yml
>    ```
> 5. Edit Keycloak to force SSL redirection
> 
> **Note**: Edit keycloak.xml, not standalone or standalone-ha
> 
> https://www.keycloak.org/docs/latest/server_installation/index.html#_setting-up-a-load-balancer-or-proxy
> 
> https://www.keycloak.org/docs/latest/server_installation/index.html#enable-https-ssl-with-a-reverse-proxy
> 
> ```
> ````bash
> sudo nano /data/dcm4chee-arc/keycloak/configuration/keycloak.xml
> ````
> ```
> 
> 1. Add the Lets Encrypt certificate to the Keycloak Truststore
>    
>    * Retrieve the certificate
>    
>    ```shell
>    kubectl get secret letsencrypt-certs -o yaml
>    ```
>    
>    
>    
>    * Copy the certificate (tls.crt) and decode it from base64:
>    
>    ```shell
>    echo 'Paste tls.crt here' | base64 --decode > certificate.crt
>    ```
>    
>    
>    ```shell
>    sudo apt install openjdk-9-jre-headless
>    
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias letsencrypt -file certificate.crt
>    
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias letsencrypt -file certificate.crt
>    
>    - Remove the decoded certificate, for safety:
>    
>    	````bash
>    	rm certificate.crt
>    	````
>    
>    - Note: If you are using Auth0, you will need to add the AmazonRootCA1 certificate as well! See https://www.amazontrust.com/repository/AmazonRootCA1.pem
>    
>    - If you are using Auth0 with a custom domain, you will need:
>    ```
>    
>    
>    curl https://letsencrypt.org/certs/isrgrootx1.pem.txt > isrgrootx1.pem
>    curl https://letsencrypt.org/certs/lets-encrypt-x3-cross-signed.pem.txt > lets-encrypt-x3-cross-signed.pem
>    curl https://letsencrypt.org/certs/letsencryptauthorityx3.pem.txt > letsencryptauthorityx3.pem
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias le1 -file isrgrootx1.pem
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias le2 -file lets-encrypt-x3-cross-signed.pem
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/keycloak/configuration/keycloak/cacerts.jks -storepass secret -noprompt -importcert -alias le3 -file letsencryptauthorityx3.pem
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias le1 -file isrgrootx1.pem
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias le2 -file lets-encrypt-x3-cross-signed.pem
>    sudo keytool -trustcacerts -keystore /data/dcm4chee-arc/wildfly/configuration/dcm4chee-arc/cacerts.jks -storepass secret -noprompt -importcert -alias le3 -file letsencryptauthorityx3.pem
>    ```
>    
>    https://groups.google.com/forum/#!searchin/dcm4che/Let$27s$20Encrypt%7Csort:date/dcm4che/rtMYHztMWxY/chINDE4RCAAJ
>    ```
> 2. Create OIDC clients for dcm4chee and the OHIF Viewer in Keycloak
> 
> **Note:** You will need to add a Valid Redirect URI for the HTTP route to dcm4chee, in addition to the HTTPs route. This is a known issue.
> 
> 1. (Optional) Setup IHE Invoke Image Display routes in dcm4chee
> 
> Notes:
> 
> * Had to disable SSL to set up Keycloak initially
> * If LE certificate needs to be renewed, you may have to add it to dcm4chee and keycloak's keystore again
> * If you run into 'no path to resource' errors it's probably an issue like "part of the SSL cert is missing in the keystore"

Thank you so much for this. I will try it out and document whether it works or not. 

It seems like a long process though, how would I go about filtering the studies retrieved from the remote server by `Patient Name` as a short term hack while I apply your steps on the side?

**Comment by omartoutounji** [comment]
For any other user who visits this issue, to filter the studies as a short term hack:

1. Go to `StudyListWithData.js`

2. In the `searchForStudies()` method, there's a `const filter` constant that you can modify to your liking and get the results you want
